### ISSUE TITLE
DL Pack with PyTorch is broken since PyTorch 1.13 due to normalised strides for dimension of size 1

### ISSUE THREAD

**Issue by MatthieuToulemont** [triton-inference-server/server#6102]
**Description**
PyTorch updated the way they handle strides for dimensions of size 1. 

In Pytorch 1.12 : 

```
import torch
from torch.utils.dlpack import to_dlpack, from_dlpack

x = torch.randn((1, 1, 768,768)).cuda()

y = from_dlpack(to_dlpack(x))

print(x.stride()) # (589824, 589824, 768, 1)
print(y.stride()) # (589824, 589824, 768, 1)
```

In PyTorch `>=` 1.13: 

```
import torch
from torch.utils.dlpack import to_dlpack, from_dlpack

x = torch.randn((1, 1, 768,768)).cuda()

y = from_dlpack(to_dlpack(x))

print(x.stride()) # (589824, 589824, 768, 1)
print(y.stride()) # (1, 1, 768, 1)
```

The code in triton to check if a DLPack tensor is contiguous is [here](https://github.com/triton-inference-server/python_backend/blob/bae80b1e49759745208fd8c0aaf0bfece0261256/src/pb_tensor.cc#L432). 
```
// Check if the input is contiguous and in C order
  if (strides != nullptr) {
    int64_t calculated_stride{1};
    bool is_contiguous_c_order = true;
    for (size_t i = 1; i < dims.size(); i++) {
      if (strides[ndim - i] != calculated_stride) {
        is_contiguous_c_order = false;
        break;
      }

      calculated_stride *= dims[ndim - i];
    }
```

Would updating this snippet of code to : 
```
// Check if the input is contiguous and in C order
  if (strides != nullptr) {
    int64_t calculated_stride{1};
    bool is_contiguous_c_order = true;
    for (size_t i = 1; i < dims.size(); i++) {
      if (dims[ndim-i] != 1){
        if (strides[ndim - i] != calculated_stride) {
          is_contiguous_c_order = false;
          break;
        }
        calculated_stride *= dims[ndim - i];
     }
    }
```

work ? Or would that break further down the line ? 

I am pretty new to C++ but would be glad to help either fixing this. 

The reason I am bringing this to Triton is that PyTorch made this change to conform to what numpy does (see this [issue](https://github.com/pytorch/pytorch/issues/99803#issuecomment-1519632449)) and it doesn't look like they would be ready to roll it back. 

This issue is rather unfortunate as it forces to stay with `PyTorch 1.12` and prevents us from benefitting from the compilation features of `PyTorch >= 2.0`

**Triton Information**
What version of Triton are you using? 23.06

Are you using the Triton container or did you build it yourself? Docker and install pytorch 2.0.1

Best, 

Matthieu

**Comment by MatthieuTPHR** [comment]
I just rebuild the python backend with this change and it fixes my issue, I will open a PR but will leave this issue open, it case the Triton team thinks it might cause further issues down the line
